### ISSUE TITLE
Opacus won't work with Syft 0.2.9.

### ISSUE THREAD

**Issue by gkaissis** [gkaissis/PriMIA#47]
- What we need to do:
Be able to run ResNet18 with Opacus on Syft 0.2.9 release (= git commit 3852ca9). 

- What happens:
On 0.2.9 release, we get an Opacus error (Parameter has no attribute grad_sample). This is caused by hooked torch lacking some kinda patch to hook into backprop.
The issue was attempted to be fixed initially by Théo in commit c83e615 which adds the functionality. However, this causes a regression with architectures like ResNet or VGG16 and only works with simple sequential models.
After c83e615, the Opacus error is never reached. Instead, we get a PySyft error: 'NoneType' object has no attribute 'child'

These are the steps to reproduce:

- Clone https://github.com/gkaissis/PriMIA.git
- git switch to the branch opacus_fix
- pip install the following in the following order:
    - git+https://github.com/OpenMined/PySyft@514c40aafa79a49fbca498484a7d7b541dddb703
    - tqdm
    - opacus==0.9.1
    - albumentations
    - pandas
    - tabulate

- make sure you are in the root of the repository
- run make minimal_server_folders
- run python revision_scripts/dp_training.py

CAUTION: If you try to install different syft versions while debugging, make sure to do the following every time:
- pip uninstall syft --yes
- pip install <whichever version or hash you install> --no-dependencies
If you don't do this, torch will get reverted to 1.4 every time. Opacus installs torch 1.6, which seems to not cause a problem anywhere else in this story.

You should get the error 'NoneType' object has no attribute 'child'
The error comes from running loss.backward() and happens before optimiser.step() (Opacus) is invoked. 
The error only happens after commit c83e615 which is Théos fix for Opacus. Before that you get an Opacus error (Parameter has no attribute grad_sample).
It seems that the c83e615 commit introduced some regression which causes ResNet to break when calculating the backward() function.

**Comment by gkaissis** [comment]
Extra piece of information. When not using ResNet, the training with Opacus starts, but the following warning is given: 

`UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the gradient for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations.`

**Comment by Darktex** [comment]
Hi guys! I'm one of the maintainers of Opacus :) Can I ask you to file an [issue](https://github.com/pytorch/opacus/issues) with a way to reproduce your problem? I'll gladly take a look :) 

(If you could do a colab that would be super appreciated because the Facebook way of running python scripts is a bit odd... But if not don't sweat it)

**Comment by Darktex** [comment]
> Extra piece of information. When not using ResNet, the training with Opacus starts, but the following warning is given:
> 
> `UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the gradient for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations.`

This would happen if you had something in the graph after loss, ie your forward graph computed something more of top of the loss. At that point, that "extra" thing would be the leaf node, and taking a gradient of the loss would return that error since it's not a leaf anymore

**Comment by BambooPalace** [comment]
This issue with Parameter has no attribute grad_sample still exists, did u manage to solve it?
